python/cudf/cudf/tests/test_joining.py

```diff
@@ -1206,6 +1207,177 @@ def test_typecast_on_join_categorical(dtype_l, dtype_r):
     assert_eq(expect, got)
 
 
+def make_categorical_dataframe(categories, ordered=False):
+    dtype = CategoricalDtype(categories=categories, ordered=ordered)
+    data = cudf.Series(categories, dtype=dtype)
+    return cudf.DataFrame({"key": data})
+
+
+def test_categorical_typecast_inner():
+    # Inner join casting rules for categoricals
+
+    # Equal categories, equal ordering -> common categorical
+    left = make_categorical_dataframe([1, 2, 3], ordered=False)
+    right = make_categorical_dataframe([1, 2, 3], ordered=False)
+    result = left.merge(right, how="inner", on="key")
+
+    expect_dtype = CategoricalDtype(categories=[1, 2, 3], ordered=False)
+    expect_data = cudf.Series([1, 2, 3], dtype=expect_dtype, name="key")
+
+    assert_eq(expect_data, result["key"])
+
+    # Equal categories, unequal ordering -> error
+    left = make_categorical_dataframe([1, 2, 3], ordered=False)
+    right = make_categorical_dataframe([1, 2, 3], ordered=True)
+
+    with pytest.raises(TypeError):
+        result = left.merge(right, how="inner", on="key")
+
+    # Unequal categories
+    # Neither ordered -> unordered categorical with intersection
+    left = make_categorical_dataframe([1, 2, 3], ordered=False)
+    right = make_categorical_dataframe([2, 3, 4], ordered=False)
+
+    result = left.merge(right, how="inner", on="key")
+
+    expect_dtype = cudf.CategoricalDtype(categories=[2, 3], ordered=False)
+    expect_data = cudf.Series([2, 3], dtype=expect_dtype, name="key")
+    assert_eq(expect_data, result["key"])
+
+    # One is ordered -> error
+    left = make_categorical_dataframe([1, 2, 3], ordered=False)
+    right = make_categorical_dataframe([2, 3, 4], ordered=True)
+
+    with pytest.raises(TypeError):
+        result = left.merge(right, how="inner", on="key")
+
+    # Both are ordered -> error
+    left = make_categorical_dataframe([1, 2, 3], ordered=True)
+    right = make_categorical_dataframe([2, 3, 4], ordered=True)
+
+    with pytest.raises(TypeError):
+        result = left.merge(right, how="inner", on="key")
+
+
+def test_categorical_typecast_left():
+    # TODO: generalize to right or write another test
+    # Inner join casting rules for categoricals
+
+    # equal categories, neither ordered -> common dtype
+    left = make_categorical_dataframe([1, 2, 3], ordered=False)
+    right = make_categorical_dataframe([1, 2, 3], ordered=False)
+
+    result = left.merge(right, on="key", how="left")
+
+    expect_dtype = CategoricalDtype(categories=[1, 2, 3], ordered=False)
+    expect_data = cudf.Series([1, 2, 3], dtype=expect_dtype, name="key")
+
+    assert_eq(expect_data, result["key"])
+
+    # equal categories, left is ordered, right is not
+    # left dtype takes precedence
```

Comment:
right - that's the idea. I'm imagining this pops up in workflows where a user has a big, important table they're working on and wants to left merge that table onto another dataset that they just read in, or something like that. It felt unfortunate to force them to order the categories of that other table before allowing a join, rather we might assume that the left table is really the source of truth. 

I'd argue the reverse case *should* raise though - left is unordered, right is ordered - because now the user either has to throw away information (the ordering from the right) or impose that information onto a column that doesn't necessarily have that same meaning.